### PR TITLE
Fix disposal of enumerators in EnumerablePartition

### PR DESCRIPTION
**Pull request by jamesqo**
**Description:** I found out while adding a new API in https://github.com/dotnet/corefx/pull/14186 that I had forgotten to override `Dispose` in `EnumerablePartition`, which is returned by `Skip` / `Take`. This fixes that and adds corresponding tests to make sure the source enumerator is disposed properly. I also made some of the test code I had written in an earlier PR a bit cleaner.

**Related:** https://github.com/dotnet/corefx/pull/13628

/cc @JonHanna @stephentoub @VSadov

### CODE REVIEW

File: src/System.Linq/tests/SelectManyTests.cs [hunk 1]
```diff
@@ -409,31 +409,19 @@ public void DisposeAfterEnumeration(int sourceLength, int subLength)
...
-                DisposeWorker = () => subCollectionDisposed[subIndex++] = true
-            };
-
-            var subCollection = new DelegateBasedEnumerable<int>
-            {
-                GetEnumeratorWorker = () => subEnumerator
-            };
+            DelegateIterator<int> source = null;
+            source = new DelegateIterator<int>(
+                getEnumerator: () => source,
```
Comment (stephentoub) [inline src/System.Linq/tests/SelectManyTests.cs]: Nit: if you were to change the implementation in the ctor to:
```C
_getEnumerator = getEnumerator ?? () => this;
```
then I think you could change this to:
```C#
var source = new DelegateIterator<int>(
    moveNext: ...
```
etc., and not have to do this dance with defining the variable separately.

**Comment by VSadov** [comment]
LGTM. Thanks!

File: src/System.Linq/tests/SelectManyTests.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.Linq/tests/SelectManyTests.cs]: Done.

**Comment by jamesqo** [comment]
Test Innerloop OSX Debug Build and Test (build timed out)
Test Innerloop Windows_NT Debug Build and Test (reported SqlClient failures)
Test Innerloop Windows_NT Release Build and Test (reported SqlClient failures)

**Pull request merged**